asapdiscovery-dataviz/asapdiscovery/dataviz/html_viz.py

```diff
@@ -570,3 +769,81 @@ def make_poses_html(self):
             html = self.get_html_airium(pose)
             html_renders.append(html)
         return html_renders
+
+    def make_logoplot_input(self, resi) -> dict:
+        """
+        given a residue number with underscored chain ID, get data for the fitness of all mutants for the residue. Use
+        LogoMaker to create a logoplot for both the fit and unfit mutants, return the base64
+        string of the image.
+        """
+
+        # get just the fitness data for the queried residue index, at the right chain.
+        resi, chain = resi.split("_")
+        site_df_resi = self.fitness_data_logoplots[
+            self.fitness_data_logoplots["site"] == int(resi)
+        ]
+        site_df = site_df_resi[site_df_resi["chain"] == chain]
+        # add the fitness threshold to normalize so that fit mutants end up in the left-hand logoplot.
+        site_df["fitness"] = site_df["fitness"] + abs(
+            _FITNESS_DATA_FIT_THRESHOLD[TargetVirusMap[self.target]]
+        )
+
+        # split the mutant data into fit/unfit.
+        site_df_fit = site_df[site_df["fitness"] > 0]
+        site_df_unfit = site_df[site_df["fitness"] < 0]
+
+        if len(site_df_fit) == 0:
+            raise ValueError(
+                f"No fit mutants found for residue {resi} in chain {chain}. Are you sure the fitness threshold is set correctly? At least the wildtype residue should be fit."
+            )
+        elif len(site_df_unfit) == 0:
+            print(
+                f"Warning: no unfit residues found for residue {resi} in chain {chain}."
+            )
+            # make a row with a fake unfit mutant instead.
+            site_df_unfit.loc[0] = [
+                site_df_fit["gene"].values[0],
+                resi,
+                "X",
+                -0.00001,
+                0,
+                site_df_fit["wildtype"].values[0],
+                chain,
+            ]
+
+        logoplot_base64s_dict = {}
+        with tempfile.TemporaryDirectory() as tmpdirname:
+            for fit_type, fitness_df in zip(
+                ["fit", "unfit"], [site_df_fit, site_df_unfit]
+            ):
+                # pivot table to make into LogoMaker format
+                logoplot_df = pd.DataFrame(
+                    [fitness_df["fitness"].values], columns=fitness_df["mutant"]
+                )
+
+                # create Logo object
+                logomaker.Logo(
+                    logoplot_df,
+                    shade_below=0.5,
+                    fade_below=0.5,
+                    font_name="Sans Serif",
+                    figsize=(3, 10),
+                    color_scheme="dmslogo_funcgroup",
+                    flip_below=False,
+                    show_spines=True,
+                )
+                plt.xticks([])
+                plt.yticks([])
+
+                # we could get base64 from buffer, but easier to write as tmp and read back as bas64.
```

Comment:
would prefer in memory if possible, but this module is full of tmpdir stuff anyway, so lets just roll with it.

Reply:
yeah it's not ideal, but I found some examples (e.g. [this one](https://stackoverflow.com/questions/66324439/matplotlib-image-to-base64-without-saving)) that would make this section less readable imo.